io/input/router.go

```diff
@@ -193,12 +185,34 @@ func (s Source) Focused(tag event.Tag) bool {
 	return s.r.state().keyState.focus == tag
 }
 
+// SourceExecuteProcessor is the default command processor for a Source,
+// usually it should NOT be overridden.
+//
+// There's few legitimate reasons to override this function, one of them is to
+// implement a custom command processor or testing.
+var SourceExecuteProcessor = (*Router).Execute
+
+// Execute a command.
+func (s Source) Execute(c Command) {
+	if !s.Enabled() {
+		return
+	}
+	SourceExecuteProcessor(s.r, c)
+}
+
+// SourceEventProcessor is the default event processor for a Source,
+// usually it should NOT be overridden.
+//
+// There's few legitimate reasons to override this function, one of them is to
+// implement a custom event processor or testing.
+var SourceEventProcessor = (*Router).Event
+
 // Event returns the next event that matches at least one of filters.
 func (s Source) Event(filters ...event.Filter) (event.Event, bool) {
 	if !s.Enabled() {
 		return nil, false
 	}
-	return s.r.Event(filters...)
+	return SourceEventProcessor(s.r, filters...)
```

Comment:
I believe this indirection will cause `filters` to escape the heap, causing an allocation for every call to `Source.Event`.

Reply:
> I believe this indirection will cause `filters` to escape the heap, causing an allocation for every call to `Source.Event`.

So a callback function method should be added?

Reply:
I also prepared a file drag and drop implementation for gio for the Windows platform, but the issue of passing events is a bit confusing, as he said, the current gui can only receive events can not generate custom events which brings confusion to the implementation of the file drag and drop events